kernel/arch/dreamcast/hardware/asic.c

```diff
@@ -230,3 +252,99 @@ void asic_init(void) {
 void asic_shutdown(void) {
     asic_evt_shutdown();
 }
+
+static void * asic_threaded_irq(void *data) {
+    struct asic_thdata *thdata = data;
+    int flags;
+
+    for (;;) {
+        flags = irq_disable();
+
+        if (!thdata->pending)
+            genwait_wait(&thdata->genwait_obj, "Threaded IRQ", 0, NULL);
+
+        irq_restore(flags);
+
+        if (thdata->quit)
+            break;
+
+        thdata->pending = false;
+        thdata->hdl(thdata->source, thdata->data);
+
+        if (thdata->unmask)
+            thdata->unmask(thdata->source);
+    }
+
+    return NULL;
+}
+
+static void asic_thirq_dispatch(uint32_t source, void *data) {
+    struct asic_thdata *thdata = data;
+
+    if (thdata->ack_and_mask)
+        thdata->ack_and_mask(source);
+
+    thdata->source = source;
+
+    thdata->pending = true;
+    genwait_wake_one(&thdata->genwait_obj);
+}
+
+int asic_request_threaded_handler(uint16_t code, asic_evt_handler hnd,
+                                  void *data,
+                                  void (*ack_and_mask)(uint16_t),
+                                  void (*unmask)(uint16_t))
+{
+    struct asic_thdata *thdata;
+    uint32_t flags;
+
+    thdata = malloc(sizeof(*thdata));
+    if (!thdata)
+        return -1; /* TODO: What return code? */
+
+    thdata->hdl = hnd;
+    thdata->data = data;
+    thdata->quit = false;
+    thdata->pending = false;
+    thdata->ack_and_mask = ack_and_mask;
+    thdata->unmask = unmask;
+
+    flags = irq_disable();
+
+    thdata->thd = thd_create(0, asic_threaded_irq, thdata);
+    if (!thdata->thd) {
+        irq_restore(flags);
+        free(thdata);
+        return -1; /* TODO: What return code? */
+    }
+
+    /* Highest priority */
+    //thd_set_prio(thdata->thd, 0);
+
+    asic_evt_set_handler(code, asic_thirq_dispatch, thdata);
+
+    irq_restore(flags);
+
+    return 0;
+}
+
+void asic_remove_threaded_handler(uint16_t code)
```

Comment:
Your `asic_remove_handler` wouldn't know which handler is a threaded interrupt and which one is not, because it only has a `void *data` pointer...

Reply:
That was meant to be in the context of the other comment to add a way to distinguish them, but you figured that out by the end :) . My thinking was that for the handlers we want to make threaded the use cases we have seen have been that we want user-installable callbacks. It would make sense in those cases to use a regular handler by default ( which would likely be something like `{ ack_and_mask_func(); unmask_func(); }` ) then if a user requests a custom callback install the threaded version.

Overly long comment is just to document that reasoning for posterity.